File: src/0210_course_schedule_ii.js
function findOrder(numCourses, prerequisites) {
    const prereq = {};
    for (let c = 0; c < numCourses; c++) prereq[c] = [];
    for (const [crs, pre] of prerequisites) prereq[crs].push(pre);
    const output = [];
    const visit = new Set();
    const cycle = new Set();
    const dfs = (crs) => {
        if (cycle.has(crs)) return false;
        if (visit.has(crs)) return true;
        cycle.add(crs);
        for (const pre of prereq[crs]) {
            if (dfs(pre) === false) return false;
        }
        cycle.delete(crs);
        visit.add(crs);
        output.push(crs);
        return true;
    };
    for (let c = 0; c < numCourses; c++) {
        if (dfs(c) === false) return [];
    }
    return output;
}

module.exports = { findOrder };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    assert.deepStrictEqual(findOrder(2, [[1, 0]]), [0, 1]);
    assert.deepStrictEqual(findOrder(4, [[1, 0], [2, 0], [3, 1], [3, 2]]), [0, 1, 2, 3]);
    assert.deepStrictEqual(findOrder(1, []), [0]);
}

File: src/0210_course_schedule_ii.rs
struct Solution;

use std::collections::HashSet;

impl Solution {
    fn course_dfs(
        crs: i32,
        prereq: &[Vec<i32>],
        output: &mut Vec<i32>,
        visit: &mut HashSet<i32>,
        cycle: &mut HashSet<i32>,
    ) -> bool {
        if cycle.contains(&crs) {
            return false;
        }
        if visit.contains(&crs) {
            return true;
        }

        cycle.insert(crs);
        for &pre in &prereq[crs as usize] {
            if !Self::course_dfs(pre, prereq, output, visit, cycle) {
                return false;
            }
        }
        cycle.remove(&crs);
        visit.insert(crs);
        output.push(crs);
        true
    }

    pub fn find_order(num_courses: i32, prerequisites: Vec<Vec<i32>>) -> Vec<i32> {
        let mut prereq = vec![Vec::new(); num_courses as usize];
        for edge in prerequisites {
            prereq[edge[0] as usize].push(edge[1]);
        }

        let mut output = Vec::new();
        let mut visit = HashSet::new();
        let mut cycle = HashSet::new();
        for c in 0..num_courses {
            if !Self::course_dfs(c, &prereq, &mut output, &mut visit, &mut cycle) {
                return Vec::new();
            }
        }
        output
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn mirrors_python_main_assertions() {
        assert_eq!(Solution::find_order(2, vec![vec![1, 0]]), vec![0, 1]);
        assert_eq!(
            Solution::find_order(4, vec![vec![1, 0], vec![2, 0], vec![3, 1], vec![3, 2]]),
            vec![0, 1, 2, 3]
        );
        assert_eq!(Solution::find_order(1, vec![]), vec![0]);
    }
}
